src/cargo/sources/registry/http_remote.rs

```diff
@@ -0,0 +1,741 @@
+            .trim_start_matches("sparse+")
+            .into_url()
+            .expect("a url with the protocol stripped should still be valid");
+
+        HttpRegistry {
+            index_path: config.registry_index_path().join(name),
+            cache_path: config.registry_cache_path().join(name),
+            source_id,
+            config,
+            url,
+            multi: Multi::new(),
+            multiplexing: false,
+            downloads: Downloads {
+                start: Instant::now(),
+                config,
+                next: 0,
+                pending: HashMap::new(),
+                pending_ids: HashMap::new(),
+                results: HashMap::new(),
+                progress: RefCell::new(Some(Progress::with_style(
+                    "Fetching",
+                    ProgressStyle::Ratio,
+                    config,
+                ))),
+                downloads_finished: 0,
+                downloaded_bytes: 0,
+                success: false,
+            },
+            fresh: HashSet::new(),
+            requested_update: false,
+        }
+    }
+
+    fn handle_http_header(buf: &[u8]) -> Option<(&[u8], &str)> {
+        if buf.is_empty() {
+            return None;
+        }
+
+        let mut parts = buf.splitn(2, |&c| c == b':');
+        let tag = parts.next().expect("first item of split is always Some");
+        let rest = parts.next()?;
+        let rest = std::str::from_utf8(rest).ok()?;
+        let rest = rest.trim();
+
+        // Don't let server sneak extra lines anywhere.
+        if rest.contains('\n') {
+            return None;
+        }
+
+        Some((tag, rest))
+    }
+
+    fn start_fetch(&mut self) -> CargoResult<()> {
+        // NOTE: lifted from src/cargo/core/package.rs
+        //
+        // We've enabled the `http2` feature of `curl` in Cargo, so treat
+        // failures here as fatal as it would indicate a build-time problem.
+        self.multiplexing = self.config.http_config()?.multiplexing.unwrap_or(true);
+
+        self.multi
+            .pipelining(false, self.multiplexing)
+            .with_context(|| "failed to enable multiplexing/pipelining in curl")?;
+
+        // let's not flood the server with connections
+        self.multi.set_max_host_connections(2)?;
+
+        Ok(())
+    }
+
+    fn handle_pending_downloads(&mut self) -> CargoResult<()> {
+        trace!("pending downloads: {}", self.downloads.pending.len());
+        assert_eq!(
+            self.downloads.pending.len(),
+            self.downloads.pending_ids.len()
+        );
+
+        // Collect the results from the Multi handle.
+        let pending = &self.downloads.pending;
+        let mut results = Vec::new();
+        self.multi.messages(|msg| {
+            let token = msg.token().expect("failed to read token");
+            let (_download, handle) = &pending[&token];
+            if let Some(result) = msg.result_for(handle) {
+                results.push((token, result));
```

Comment:
Why add it to a `Vec` and then iterate over it? Would it be clearer to handle each on in the callback?

Reply:
Originally I did it because the error handling was difficult inside the callback, but it turns out there really wasn't much. Fixed.